new Feature: order Apps by Usage

thereiskeks opened the pull request:
I've added a new Comparator that sorts Apps by usage. I also added the attribute **usagesQuantity** to **LaunchableActivity** and **LaunchableActivityPrefs**. I hope everything in **LaunchableActivityPrefs** is correct. I increased **DATABASE_VERSION** to 3. Please check if thats correct or if I need to improve something.

I added the **pref_app_preferred_order_entries_usages** key in **strings** file and the german translation.

I've tested the Feature and everything works fine. I've rebooted the device to see if saving to DB works and it works.

seizonsenryaku commented on app/src/main/java/com/hayaisoftware/launcher/comparators/UsageOrder.java:
```diff
@@ -0,0 +1,38 @@
+        int lhsUsageQuantity = lhs.getusagesQuantity();
+        int rhsUsageQuantity = rhs.getusagesQuantity();
+        
+        if (lhsUsageQuantity > rhsUsageQuantity)
```
You can simplify the code to remove the "if"s.
Replace by "return rhsUsageQuantity - lhsUsageQuantity"

seizonsenryaku commented on app/src/main/java/com/hayaisoftware/launcher/activities/SearchActivity.java:
```diff
@@ -490,9 +507,12 @@ private void updateVisibleApps() {
     }
 
     private void sortApps() {
-        Collections.sort(mActivityInfos, mAlphabeticalOrderComparator);
```
Keep this line. We always order alphabetically first, so in case of a "tie" in the selected comparison setting the app order stays alphabetical.

seizonsenryaku commented on app/src/main/java/com/hayaisoftware/launcher/activities/SearchActivity.java:
```diff
@@ -490,9 +507,12 @@ private void updateVisibleApps() {
+        } else if(mShouldOrderByUsages) {
+            Collections.sort(mActivityInfos, mUsageOrderComparator);
+        } else {
+            Collections.sort(mActivityInfos, mAlphabeticalOrderComparator);
```
Remove this.

seizonsenryaku commented:
Thank you for contributing! Can you address my comments?

thereiskeks commented on app/src/main/java/com/hayaisoftware/launcher/activities/SearchActivity.java:
```diff
@@ -507,13 +507,14 @@ private void updateVisibleApps() {
             Collections.sort(mActivityInfos, mRecentOrderComparator);
         } else if(mShouldOrderByUsages) {
             Collections.sort(mActivityInfos, mUsageOrderComparator);
-        } else {
```
I only removed the trailing else and put the alphabetical sorting above the if. Should work as expected now.

thereiskeks commented:
I've addressed your issues. How will the missing languages be approached? It defaults to english at the moment and I've added the word to the German translation.

seizonsenryaku commented:
I usually release with missing translations and people contribute afterwards.
Is the "com.google.android.gms:play-services-appindexing:8.1.0" actually used for anything?

thereiskeks commented:
I think it hast something to do with [App Indexing Api](https://developers.google.com/app-indexing/android/publish). I haven't added it intentionally, Android Studio must have added it automatically. I'm a bit ashamed of committing something that doesn't seem to be relevant :D.

seizonsenryaku commented:
Sorry about the huge delay in merging this =/

seizonsenryaku commented:
This actually caused a crash on updating from a previous version because of a missing column in tbe database. The correct fix would be only adding a new column in the onUpgrade, but I realized the error after updating to the play store, so I had to do the fastest fix I could code (dropping the previous version of the database...).
